### ISSUE TITLE
Add missing functionalities to support API

### ISSUE THREAD

**Issue by cojoj** [czechboy0/XcodeServerSDK#40]
Apple has left as with a table of **supported** endpoint in Xcode 7. It shows up that they're only providing some basic stuff but it's necessary for us to cover them all!
We're on a good way and already cover most but some of them are missing - not because we don't have code to support them but because a nice, convenience method hasn't beed written yet (you can do it).

In `README` you can see a [table of features](https://github.com/czechboy0/XcodeServerSDK#features), so if you're planning to implement some of those _missing_ methods, plase update `README` as well

---

Here's the list of all _official_ endpoints:
-  [x] List bots on server
- [x] Create a new bot
- [x] Retrieve a bot by ID
- [ ] Update a bot’s configuration (**will be tricky**)
- [x] Get the most recent integrations for a bot
- [x] Enqueue a new integration 
- [x] List integrations on server 
- [x] Retrieve an integration by ID
- [x] List the commits included in an integration
- [x] List the build issues produced by an integration 
- [x] List devices connected to server
- [x] List hosted repositories on server
- [x] Create a new hosted repository

---

> Should be an easy one for anyone with 20mins on their hands.
> -- @czechboy0 in #39

---

If you're wondering how to determine what endpoint will be used in official API, here's the list:

| Type | Path | Description |
| --- | --- | --- |
| GET | `/bots` | List bots on server |
| POST | `/bots` | Create a new bot |
| GET | `/bots/(id)` | Retrieve a bot by ID |
| PATCH | `/bots/(id)` | Update a bot’s configuration |
| GET | `/bots/(id)/integrations` | Get the most recent integrations for a bot |
| POST | `/bots/(id)/integrations` | Enqueue a new integration |
| GET | `/integrations` | List integrations on server |
| GET | `/integrations/(id)` | Retrieve an integration by ID |
| GET | `/integrations/(id)/commits` | List the commits included in an integration |
| GET | `/integrations/(id)/issues` | List the build issues produced by an integration |
| GET | `/devices` | List devices connected to server |
| GET | `/repositories` | List hosted repositories on server |
| POST | `/repositories` | Create a new hosted repository |

If you need some _motivation_ to start doing this task, please refer to #2 😉

**Comment by czechboy0** [comment]
:+1:

**Comment by cojoj** [comment]
I think I'll cover this if only I get **OS X Server** running on my MBP 😡

**Comment by cojoj** [comment]
@czechboy0, @esttorhe I'm trying to update `XcodeServerSDK` to support missing API calls but I'm stuck on the first one... 😁

**Update a bot’s configuration** - how do you imagine the method signature for this one? I came with two ideas but don't know which one will be better:
- `public func updateBot(botId: String, withBot bot: Bot, overwriteBlueprint blueprint: Bool = true, completion: (success: Bool, error: NSError?) -> ())`
- `public func updateBot(bot: Bot, overwriteBlueprint blueprint: Bool = true, completion: (success: Bool, error: NSError?) -> ())`

In the first option you specify both: **bot to update** by ID and **updating bot**. In the latter you just provide updated bot with correct ID which will be used to `PATCH` existing bot. Which one do you prefer more?

**Comment by czechboy0** [comment]
Hmm this one is actually pretty tricky. I suggest we leave that one for last :laughing:

**Comment by czechboy0** [comment]
Also because some of the work I'm doing now will make the bot updating call much easier to work with. If you can take a look at `List the build issues produced by an integration`, `List the commits included in an integration`, `Retrieve an integration by ID` etc that'd be most useful.

**Comment by cojoj** [comment]
Typical me - start working from the _worst_ side 😐
Roger that, I will go for other things and leave this for later.

Thing that get stuck in my head after opening `XcodeServer` class - why not move those calls to separate files as `extension`s eg. `XcodeServer+Bot`, `XcodeServer+Integrations`. This way we'll only have generic `sendRequestWithMethod()` in XcodeServer class and get rid specific API calls. But I guess, we'll have to wait with this until your PR as this'll some introduce merge conflicts...

**Comment by czechboy0** [comment]
Yeah I agree, now that there are more API calls it'd make sense to start splitting it. As you said, let's wait for that until this is merged (and even create an issue so that we don't forget).

**Comment by cojoj** [comment]
**Retrieve an integration by ID** covered in #48.

**Comment by cojoj** [comment]
@czechboy0 you know anything about integration commits? I have configured bot and if I run integration I always see that there are no **commits** related to this integration. Any idea when those commits are assigned to the integration?

**Comment by czechboy0** [comment]
Yeah that's going to be a tricky one to test. The commits of an integration are the commits that have been pushed **since the last integration**. So I'd suggest you create a dummy branch, run an integration there once. Then, commit a couple of times there and then run an integration again. The second one should show all those commits. However, for debugging, you can also **delete** integrations (not just bots) through Xcode. This might help in development.

**Comment by cojoj** [comment]
You're probably familiar with the output from `/integrations/(id)/commits`. What do you think should be returned to the end user? Only an array of `commit`s? Because there are many, many other things...

![screen shot 2015-06-28 at 19 02 53](https://cloud.githubusercontent.com/assets/4522407/8397334/597e6f0a-1dc8-11e5-8976-0792e3d2de98.png)

**Comment by cojoj** [comment]
Hmm, I guess I found answer... We're returning an array of commiters... But we probably don't have models for this, right? Same thing with issues and repositories, I guess...

**Comment by czechboy0** [comment]
Nope, that's all yours :wink: Create model classes for `Commit`s and committers and for whatever else makes sense.

**Comment by cojoj** [comment]
Looks like it's a job for more than 20 min 😜

**Comment by czechboy0** [comment]
Yeah that might have been an underestimate :laughing:

**Comment by cojoj** [comment]
So if #44 is already merged, are we good to go with **Updating a bot’s configuration**? I'd prefer to finish non-new-stuff-required calls first and than jump on those more involving. Please, think about the **method signature** and let me know 😉

**Comment by czechboy0** [comment]
Actually, `PATCH`ing an existing bot is a bit more involved. If I were you I'd leave that for last. It's not that common of a call anyway (I never had it in Buildasaur). But I'll take a look at it next week some time.

**Comment by cojoj** [comment]
Ok, as you wish 👌

**Comment by cojoj** [comment]
**Repositories** are represented by a simple JSON:

``` json
{
      "readAccessExternalIDs": [],
      "writeAccessExternalIDs": [
        "FDF283F5-B9C3-4B43-9000-EF6A54934D4E",
        "ABCDEFAB-CDEF-ABCD-EFAB-CDEF00000050"
      ],
      "name": "Test2",
      "posixPermissions": 0,
      "httpAccessType": 1
}
```

There's no similarity between **Repository** and `XcodeServerEntity`. So, my question is, should we anyway create it as `XcodeServerEntity` child or maybe create a simple `struct`?

**Comment by czechboy0** [comment]
I'll leave that to your judgement. I don't have a strong preference either way.

**Comment by czechboy0** [comment]
Yay only 3 more API calls to support them all! The last two should be relatively easy, it's just work adding those model classes. And the Bot patching one I'll try to look at in the next couple of days.

**Comment by cojoj** [comment]
One last to go...

**Comment by Sajjon** [comment]
@czechboy0 what is the status of `Update a bot’s configuration (will be tricky)`?

Why is it tricky? And how tricky is it? this would be a great feature! since checking out code and building a project takes over 90% for each bot.  Running the unit tests takes less than 10%, that is a big overhead! 

I have btw created a Swift 3 branch for XcodeServerSDK and each pod it is dependent on: https://github.com/Sajjon/XcodeServerSDK
Should I create a PR maybe?

**Comment by Sajjon** [comment]
@cojoj If I would try developing a Edit existing Bot feature, any ideas where to start? What are the primary challenges?

**Comment by cojoj** [comment]
Whoa, it was a long time ago... And to be clear:

![](http://s2.quickmeme.com/img/26/264476d2588c8a3a8c6db005251434f2c63c7eaf5402752918719efcb70924bf.jpg)

---

How I see it being _tricky_... You have to provide correct things to **update** bot. I guess the only required thing would be `id`, but I may be wrong (this needs to be checked). Also, updating sometimes can involve this **multi stepping**, so maybe some sort of builder for `Bot`.
